treq/test/test_testing.py

```diff
@@ -0,0 +1,323 @@
+                            "Partial value overlap matches")
+        self.assertNotEqual(HasHeaders({'one': ['two', 'three']}),
+                            {'one': ['two']},
+                            "Missing value matches")
+        self.assertNotEqual(HasHeaders({'one': ['two', 'three']}),
+                            {'ten': ['six']},
+                            "Complete inequality matches")
+
+    def test_case_insensitive_keys(self):
+        """
+        The :obj:`HasHeaders` equality function ignores the case of the header
+        keys.
+        """
+        self.assertEqual(HasHeaders({'A': ['1'], 'b': ['2']}),
+                         {'a': ['1'], 'B': ['2']})
+
+    def test_case_sensitive_values(self):
+        """
+        The :obj:`HasHeaders` equality function does care about the case of
+        the header value.
+        """
+        self.assertNotEqual(HasHeaders({'a': ['a']}), {'a': ['A']})
+
+    def test_repr(self):
+        """
+        :obj:`HasHeaders` returns a nice string repr.
+        """
+        self.assertEqual("HasHeaders({'a': ['b']})",
+                         repr(HasHeaders({'A': ['b']})))
+
+
+class StringStubbingTests(TestCase):
+    """
+    Tests for :obj:`StringStubbingResource`.
+    """
+    def _get_response_stub(self, expected_args, response):
+        """
+        Make a :obj:`IStringResponseStubs` that checks the expected args and
+        returns the given response.
+        """
+        method, url, params, headers, data = expected_args
+
+        @implementer(IStringResponseStubs)
+        class Stubber(object):
+            def get_response_for(_, _method, _url, _params, _headers, _data):
+                self.assertEqual((method, url, params, data),
+                                 (_method, _url, _params, _data))
+                self.assertEqual(HasHeaders(headers), _headers)
+                return response
+
+        return Stubber()
+
+    def test_interacts_successfully_with_istub(self):
+        """
+        The :obj:`IStringResponseStubs` is passed the correct parameters with
+        which to evaluate the response, and the response is returned.
+        """
+        resource = StringStubbingResource(self._get_response_stub(
+            ('DELETE', 'http://what/a/thing', {'page': ['1']},
+             {'x-header': ['eh']}, 'datastr'),
+            (418, {'x-response': 'responseheader'}, 'response body')))
+
+        stub = StubTreq(resource)
+
+        d = stub.delete('http://what/a/thing', headers={'x-header': 'eh'},
+                        params={'page': '1'}, data='datastr')
+        resp = self.successResultOf(d)
+        self.assertEqual(418, resp.code)
+        self.assertEqual(['responseheader'],
+                         resp.headers.getRawHeaders('x-response'))
+        self.assertEqual('response body',
+                         self.successResultOf(stub.content(resp)))
+
+
+class SequenceStringStubsTests(TestCase):
```

Comment:
Note to reviewers:  I plan to write docs in a later PR, because don't want to make this PR already bigger than the monster it already is.  But the API should look something like how these tests are written - maybe we can provide a convenience API like `string_stub_treq([((expected request args), (expected response))])` instead of having to construct `StubTreq(StringStubbingResource(SequenceStringStubs([...])))`?